scripts/parallel_run_tests.py

```diff
@@ -30,20 +42,50 @@ def run_test(test_binary):
         future_to_binary = {executor.submit(run_test, binary): binary for binary in binaries}
         for future in as_completed(future_to_binary):
             completed += 1
-            binary = os.path.basename(future_to_binary[future])
+            binary_full_name = future_to_binary[future]
+            binary = os.path.basename(binary_full_name)
             result = future.result()
-            print(f'========= test binary {binary}')
-            print(f'========= stdout of {binary}:')
-            print(result[1])
-            print(f'========= stderr of {binary}:')
-            print(result[2])
             if result[0] != 0:
-                fails.append(f'========= test binary {binary} failed, exit code {result[0]}')
+                fails.append((binary_full_name, result))
+                # fails.append(f'========= test binary {binary} failed, exit code {result[0]}')
+            else:
+                show_test_result(binary, result)
 
     print(f"========= finished run {completed} test binaries")
     if fails:
-        for f in fails:
-            print(f)
-        exit(1)
+        if len(fails) <= RERUN_THRESHOLD:
```

Comment:
I feel like rerun 5 times is a bit too much. If it just passes 1 out of 5 times we should not consider it as passing.

Reply:
It's not rerun 5 times. it's allow at most five tests faild. Then retry these failed test once without parallel execution. If this time pass => pass. if still not pass => fail.

Reply:
5 is based on historical result, if all test pass locally there are 0-3 test timeout in ci. If i run every of these timeout test seperately on ci they're supposed to pass